onnxscript/function_libs/torch_lib/ops/core.py

```diff
@@ -3026,10 +3028,54 @@ def aten_imag(self: TensorType) -> TensorType:
     raise NotImplementedError()
 
 
-def aten_index(self: TensorType, indices: Optional[Sequence[TensorType]]) -> TensorType:
+def are_consecutive(lst):
+    if len(lst) == 0:
+        return True
+    else:
+        return sorted(lst) == list(range(min(lst), max(lst) + 1))
+
+
+def _is_none_in_middle(indices):
+    not_none_indices = [i for i, idx in enumerate(indices) if idx is not None]
+    return not are_consecutive(not_none_indices)
+
+
+@torch_op("aten::index", trace_only=True)
+def aten_index(self: TensorType, indices: Sequence[Optional[INT64]]) -> TensorType:
     """index.Tensor(Tensor self, Tensor?[] indices) -> Tensor"""
 
-    raise NotImplementedError()
+    ordered_indices = sorted(range(len(indices)), key=lambda i: (indices[i] is None, i))
+    ordered_indices += list(range(len(ordered_indices), len(self.shape)))
+
+    self = op.Transpose(self, perm=ordered_indices)
+
+    # Need broadcast concat lol.
+    not_none_indices = [idx for idx in indices if idx is not None]
+    broadcast_shape = np.broadcast_shapes(*[idx.shape for idx in not_none_indices])
+    final_index = op.Concat(
+        *(op.Unsqueeze(op.Expand(idx, broadcast_shape), -1) for idx in not_none_indices),
+        axis=-1,
+    )
+
+    self = op.GatherND(self, final_index, batch_dims=0)
+    if _is_none_in_middle(indices):
+        return self
+
+    # Need to transpose back.
+    adv_res_rank = len(broadcast_shape)
+    # [adv1, adv2, ..., adv_res_rank, x1, x2, ..., xk, ..., xn] -> [x1, x2, ..., xk, adv1, ..., adv_res_rank, ..., xn]
+    perm = (
+        list(range(adv_res_rank, adv_res_rank + ordered_indices[0]))
```

Comment:
I would unpack instead

Reply:
feel free to make changes. the current version is quick hack style